internal/handlers/common/aggregations/aggregations.go

```diff
@@ -22,8 +22,29 @@ import (
 	"github.com/FerretDB/FerretDB/internal/util/iterator"
 )
 
-// Stage is a common interface for all aggregation stages.
-type Stage interface {
+// ProcessorStage is a common interface for aggregation stages that process document iterator.
+type ProcessorStage interface {
 	// Process applies an aggregate stage on documents from iterator.
 	Process(ctx context.Context, iter types.DocumentsIterator, closer *iterator.MultiCloser) (types.DocumentsIterator, error)
 }
+
+// ProducerStage is a common interface aggregation stages produce documentiterator.
+type ProducerStage interface {
+	// Produce applies an aggregate stage on documents from iterator.
+	Produce(ctx context.Context, closer *iterator.MultiCloser) (types.DocumentsIterator, error)
+}
+
+// ProducerStageDataSource is a common interface for fetching from database.
+type ProducerStageDataSource interface {
+	// CollStats fetches collection statistics from the database.
+	CollStats(ctx context.Context, closer *iterator.MultiCloser) (*CollStatsResult, error)
```

Comment:
That should be `ProducerStage` that returns an iterator with one document

Reply:
So `$collStats` is a `ProducerStage` and wants to fetch data from DB and this interface allows access to DB. Maybe I misunderstood and `ProducerStage` should be implemented on pg layer? :thinking:

Reply:
You are mixing stages and what this code called Aggregation before and calls data source now.

Data source allows one to run queries.

ProducerStage produces documents. It is not important for the interface how those documents are produced – are they fetched from the database or just made up.

There are two implementations of the produce ProducerStage. Both would use a data source. The first one would return all documents from the collection. The second one would produce collection stats. (in the future, `$match` stage could implement both `ProcessorStage` and `ProducerStage` for pushdown).

Reply:
So hopefully I get this. We want to keep `ProducerStageDataSource` or `DataSource` to its `ProducerStage`. No need to expose it here.

And I'm missing `ProducerStage` that returns all documents :sweat_smile:

Reply:
There is quite a bit more cleanup I can do, but overall I like it. :hugs: